### PR TITLE
Add support for custom default values

### PR DESCRIPTION
**Pull request by mtdowling**
This commit updates Smithy IDL v2 to support custom default values
rather than only default zero values.

We didn't previoulsy support custom default values because we wanted to
be able to change default values if ever needed, we wanted to avoid
information disclosure of unreleased internal members, and we wanted
clients to be able to fill in a default zero value if a required member
is missing on the wire (implying that the member transitioned from
required to default).

While default zero values are a big simplification for clients, they do
come with awkward tradeoffs. For example, the zero value of an enum was
`""` (or it could have been the first enum in the shape definition but
that's problematic due to model filtering, and it requires a lot of
up-front planning from service teams). We also observed that there are
likely 1,000+ members already used in AWS that have default values that
are only captured through documentation. If we could actually model
defaults, it would be easier to catch changes to default values in
automated diff tools and discuss them in API reviews.

To address these concerns and support custom default values, we
recommend the following:

* clients implement presence tracking and only serialize default values if
  the member is explicitly set to the default value. This allows services
  to change defaults if ever necessary.
* servers always serialize default values unless the member is marked with
  the `@internal` trait. This makes it explicit as to what the server
  thinks the default value is while still preventing unintended
  information disclosure of unreleased features.
* To avoid downtime and account for misconfigured servers that do not
  serialize required or default members, clients attempt to populate the
  member with a default zero value.

More details and specifics can be found in
defaults-and-model-evolution.md.

Enum changes:
* Given the addition of syntax sugar for assigning values to
  defaulted structure members, this commit also allows syntactic sugar for
  assigning values to enum and intEnum shapes without the use of the
  `@enumValue` trait.
* Now that there is no default zero value for enums, the `@enumDefault`
  trait has been removed.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: designs/defaults-and-model-evolution.md [hunk 1]
```diff
@@ -0,0 +1,770 @@
...
+
+The following example defines a structure with a "language" member that has a
+default value:
+
+```
+structure Message {
+    @required
+    title: String
+
+    language: Language = "en"
```
Comment (JordonPhillips) [inline designs/defaults-and-model-evolution.md:226]: Is there validation that asserts this is a real value?

File: designs/defaults-and-model-evolution.md [hunk 2]
```diff
@@ -0,0 +1,770 @@
...
+of a member because they are using different versions of the same model.
+
+
+### Default value serialization and deserialization
+
+When deserializing a structure, a reader MUST set the member to its default
+value if the member is missing. Outside of deserialization, there MUST NOT be
+any discernible difference between an explicitly serialized member or a member
+that was assigned the default value. If such a distinction is needed, then the
+`@default` trait is inappropriate.
```
Comment (JordonPhillips) [inline designs/defaults-and-model-evolution.md:341]: You say that it should only be important for deser here, but a few sections down you talk about how it's needed for ser too.

File: designs/enum-shapes.md [hunk 3]
```diff
@@ -49,60 +49,31 @@ enum Suit {
...
-the enum defines a member with a value of "".
-
-```
-structure GetFooInput {
-    @default
-    suit: Suit
+    DIAMOND = "diamond"
+    CLUB = "club"
+    HEART = "heart"
+    SPADE = "spade"
```
Comment (JordonPhillips) [inline designs/enum-shapes.md:60]: Did you mean to omit the values here? The shape below that this being compared to is identical

**Review by JordonPhillips — CHANGES_REQUESTED** [review]

File: designs/defaults-and-model-evolution.md (hunk 1 above)
Comment (mtdowling) [reply designs/defaults-and-model-evolution.md:226]: Yes, it's built-in via NodeValidationVisitor

File: designs/enum-shapes.md (hunk 3 above)
Comment (mtdowling) [reply designs/enum-shapes.md:60]: Whoops. Meant to add the `@enumValue` trait.

File: designs/defaults-and-model-evolution.md [hunk 4]
```diff
@@ -0,0 +1,770 @@
...
+If a mis-configured server fails to serialize a value for a required member,
+to avoid downtime, clients MAY attempt to fill in an appropriate default value
+for the member:
+
+* `@default`: If the member has the `@default` trait, use its value
+* boolean: false
+* numbers: 0
+* timestamp: 0 seconds since the Unix epoch
+* string and blob: ""
+* document: null
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply designs/defaults-and-model-evolution.md:381]: Adding some nuance

File: designs/defaults-and-model-evolution.md [hunk 5]
```diff
@@ -0,0 +1,770 @@
...
+values to remove any ambiguity about the value of the most up to default
+value. When persisting data structures, regardless of if they are persisted by
+a client or server, default values SHOULD be persisted to ensure that model
+changes do not impact previously serialized values.
+
+
+#### `@default` and `@required`
+
+A member that is marked as both `@required` and `@default` indicates that
+the member MUST be serialized even if the member is set to the default value.
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply designs/defaults-and-model-evolution.md:353]: How so? The next section says:
>To allow default values to be controlled by servers, clients SHOULD NOT
serialize default values unless the member is marked as `@required` or if
the value of the member is explicitly set by the client to the default value.

File: designs/defaults-and-model-evolution.md (hunk 2 above)
Comment (mtdowling) [reply designs/defaults-and-model-evolution.md:341]: Yeah fair. I'll clean this up and simplify. I'll remove the `@default` + `@required` combo for now too so it's simpler language. We don't have a use case for that yet.

File: designs/defaults-and-model-evolution.md [hunk 6]
```diff
@@ -0,0 +1,770 @@
...
+be compatible with the shape targeted by the member and any applied constraint
+traits (for example, values for numeric types MUST be numbers that fit within
+the targeted type and match any `@length` constraints, string types match any
+`@pattern` or `@length` traits, etc).
+
+The following shapes have restrictions on their default values:
+
+* enum: can be set to any valid string _value_ of the enum.
+* intEnum: can be set to any valid integer _value_ of the enum.
+* document: can be set to `true`, `false`, string, empty list, or empty map.
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply designs/defaults-and-model-evolution.md:276]: I was thinking some implementations don't implement JSON document types as an ADT, so allowing null essentially means even required document types are nullable... but languages that wouldn't use an ADT probably have implicitly nullable types anways. So I can add null here.

**Review by JordonPhillips — APPROVED** [review]

**Review by kstich — CHANGES_REQUESTED** [review]

File: designs/defaults-and-model-evolution.md [hunk 7]
```diff
@@ -0,0 +1,756 @@
...
+shape targeted by the member and any applied constraint traits (for example,
+values for numeric types MUST be numbers that fit within the targeted type and
+match any `@length` constraints, string types match any `@pattern` or `@length`
+traits, etc).
+
+The following shapes have restrictions on their default values:
+
+* enum: can be set to any valid string _value_ of the enum.
+* intEnum: can be set to any valid integer _value_ of the enum.
+* document: can be set to `null`, `true`, `false`, string, an empty list, or
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply designs/defaults-and-model-evolution.md:278]: Yes! Added

File: docs/source/1.0/spec/core/traits/type-refinement-traits.rst.template [hunk 8]
```diff
@@ -19,131 +19,106 @@ the type of a shape.
...
+:ref:`length <length-trait>` or :ref:`pattern <pattern-trait>` traits, etc).
 
-.. _default-values:
+The following shapes have restrictions on their default values:
 
-Default zero values
--------------------
+* enum: can be set to any valid string *value* of the enum.
+* intEnum: can be set to any valid integer *value* of the enum.
+* document: can be set to ``null``, ```true``, ``false``, string, an empty
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply docs/source/1.0/spec/core/traits/type-refinement-traits.rst.template:81]: yes, fixed

File: smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java [hunk 9]
```diff
@@ -482,10 +490,107 @@ private void parseCollection(ShapeId id, SourceLocation location, CollectionShap
...
+                if (number.isFloatingPointNumber()) {
+                    throw ModelSyntaxException.builder()
+                            .shapeId(id)
+                            .message("intEnum shapes do not support floating point values")
+                            .sourceLocation(value)
+                            .build();
+                }
+                return EnumValueTrait.builder()
+                        .sourceLocation(value.getSourceLocation())
+                        .intValue(number.getValue().intValue())
```
[the remark this answers is not in the document]
Comment (mtdowling) [reply smithy-model/src/main/java/software/amazon/smithy/model/loader/IdlModelParser.java:527]: Ah sure. Added a check.

**Review by kstich — APPROVED** [review]

**Pull request merged**